File: crates/gitron-core/src/git/worktree.rs
use super::cli::run_git;
use super::error::{GitError, GitResult};
use super::types::*;

/// List all worktrees for the repository at `workdir`.
///
/// Uses `git worktree list --porcelain` which outputs blocks like:
///   worktree /path/to/worktree
///   HEAD abc123...
///   branch refs/heads/feature-a
///   <blank line>
///
/// The main worktree is always listed first.
pub fn list_worktrees(workdir: &str) -> GitResult<Vec<WorktreeInfo>> {
    let output = run_git(workdir, &["worktree", "list", "--porcelain"])?;
    Ok(parse_worktree_list(&output.stdout))
}

/// Create a new worktree.
///
/// - `path`: Where to create the worktree (absolute or relative to workdir)
/// - `branch`: Branch to check out. If None, creates a detached HEAD at HEAD.
/// - `new_branch`: If true, creates a new branch with the given name.
///
/// Maps to: `git worktree add [-b <branch>] <path> [<commit-ish>]`
pub fn add_worktree(
    workdir: &str,
    path: &str,
    branch: Option<&str>,
    new_branch: bool,
) -> GitResult<WorktreeCreateResult> {
    let mut args = vec!["worktree", "add"];

    if let Some(b) = branch {
        if new_branch {
            args.extend(&["-b", b, path]);
        } else {
            args.extend(&[path, b]);
        }
    } else {
        args.extend(&["--detach", path]);
    }

    let output = run_git(workdir, &args)?;

    // Find the newly created worktree by matching path
    let worktrees = list_worktrees(workdir)?;
    let abs_path = std::path::Path::new(workdir).join(path);
    let canonical = abs_path
        .canonicalize()
        .unwrap_or_else(|_| abs_path.to_path_buf());
    let canonical_str = canonical.to_string_lossy().to_string();

    let worktree = worktrees
        .into_iter()
        .find(|w| w.path == canonical_str || w.path == path)
        .ok_or_else(|| GitError::Other("Worktree created but not found in list".into()))?;

    Ok(WorktreeCreateResult {
        worktree,
        output: OperationOutput {
            stdout: output.stdout,
            stderr: output.stderr,
        },
    })
}

/// Remove a worktree.
///
/// - `force`: If true, removes even if there are uncommitted changes.
///
/// Maps to: `git worktree remove [--force] <worktree>`
pub fn remove_worktree(
    workdir: &str,
    worktree_path: &str,
    force: bool,
) -> GitResult<WorktreeRemoveResult> {
    let mut args = vec!["worktree", "remove"];
    if force {
        args.push("--force");
    }
    args.push(worktree_path);

    let result = run_git(workdir, &args);
    match result {
        Ok(output) => Ok(WorktreeRemoveResult {
            success: true,
            output: OperationOutput {
                stdout: output.stdout,
                stderr: output.stderr,
            },
        }),
        Err(GitError::CliError {
            stderr,
            exit_code,
            ..
        }) => Ok(WorktreeRemoveResult {
            success: false,
            output: OperationOutput {
                stdout: String::new(),
                stderr: format!("exit code {exit_code}: {stderr}"),
            },
        }),
        Err(e) => Err(e),
    }
}

/// Lock a worktree to prevent pruning.
///
/// Maps to: `git worktree lock [--reason <reason>] <worktree>`
pub fn lock_worktree(
    workdir: &str,
    worktree_path: &str,
    reason: Option<&str>,
) -> GitResult<()> {
    let mut args = vec!["worktree", "lock"];
    if let Some(r) = reason {
        args.extend(&["--reason", r]);
    }
    args.push(worktree_path);
    run_git(workdir, &args)?;
    Ok(())
}

/// Unlock a worktree.
///
/// Maps to: `git worktree unlock <worktree>`
pub fn unlock_worktree(workdir: &str, worktree_path: &str) -> GitResult<()> {
    run_git(workdir, &["worktree", "unlock", worktree_path])?;
    Ok(())
}

/// Prune stale worktree references.
///
/// Maps to: `git worktree prune [--dry-run] --verbose`
pub fn prune_worktrees(workdir: &str, dry_run: bool) -> GitResult<WorktreePruneResult> {
    let mut args = vec!["worktree", "prune", "--verbose"];
    if dry_run {
        args.push("--dry-run");
    }
    let output = run_git(workdir, &args)?;
    let pruned = output
        .stdout
        .lines()
        .filter(|l| !l.is_empty())
        .map(|l| l.to_string())
        .collect();

    Ok(WorktreePruneResult {
        pruned,
        output: OperationOutput {
            stdout: output.stdout,
            stderr: output.stderr,
        },
    })
}

/// Parse the porcelain output of `git worktree list --porcelain`.
///
/// Format (blocks separated by blank lines):
///   worktree /absolute/path
///   HEAD <oid>
///   branch refs/heads/<name>   (or "detached" for detached HEAD)
///   locked [<reason>]          (only if locked)
///   prunable                   (only if prunable)
fn parse_worktree_list(output: &str) -> Vec<WorktreeInfo> {
    let mut worktrees = Vec::new();
    let mut is_first = true;

    for block in output.split("\n\n") {
        let block = block.trim();
        if block.is_empty() {
            continue;
        }

        let mut path = String::new();
        let mut head_oid = None;
        let mut branch = None;
        let mut is_locked = false;
        let mut lock_reason = None;

        for line in block.lines() {
            if let Some(p) = line.strip_prefix("worktree ") {
                path = p.to_string();
            } else if let Some(h) = line.strip_prefix("HEAD ") {
                head_oid = Some(h.to_string());
            } else if let Some(b) = line.strip_prefix("branch ") {
                branch = Some(
                    b.strip_prefix("refs/heads/")
                        .unwrap_or(b)
                        .to_string(),
                );
            } else if line == "detached" {
                branch = None;
            } else if line == "locked" {
                is_locked = true;
            } else if let Some(reason) = line.strip_prefix("locked ") {
                is_locked = true;
                lock_reason = Some(reason.to_string());
            }
        }

        if path.is_empty() {
            continue;
        }

        let is_valid = std::path::Path::new(&path).exists();
        let head_short_oid = head_oid
            .as_ref()
            .map(|h| h.chars().take(7).collect::<String>());
        let name = if is_first {
            "main".to_string()
        } else {
            std::path::Path::new(&path)
                .file_name()
                .map(|n| n.to_string_lossy().to_string())
                .unwrap_or_else(|| path.clone())
        };

        worktrees.push(WorktreeInfo {
            name,
            path,
            branch,
            head_oid,
            head_short_oid,
            is_main: is_first,
            is_locked,
            lock_reason,
            is_valid,
        });

        is_first = false;
    }

    worktrees
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_worktree_list_single() {
        let output = "worktree /home/user/repo\nHEAD abc1234567890\nbranch refs/heads/main\n\n";
        let result = parse_worktree_list(output);
        assert_eq!(result.len(), 1);
        assert_eq!(result[0].name, "main");
        assert_eq!(result[0].path, "/home/user/repo");
        assert_eq!(result[0].branch, Some("main".to_string()));
        assert_eq!(
            result[0].head_oid,
            Some("abc1234567890".to_string())
        );
        assert_eq!(
            result[0].head_short_oid,
            Some("abc1234".to_string())
        );
        assert!(result[0].is_main);
        assert!(!result[0].is_locked);
    }

    #[test]
    fn test_parse_worktree_list_multiple() {
        let output = "\
worktree /home/user/repo
HEAD abc1234567890
branch refs/heads/main

worktree /home/user/repo-feature
HEAD def5678901234
branch refs/heads/feat/auth

";
        let result = parse_worktree_list(output);
        assert_eq!(result.len(), 2);
        assert!(result[0].is_main);
        assert_eq!(result[0].branch, Some("main".to_string()));
        assert!(!result[1].is_main);
        assert_eq!(result[1].name, "repo-feature");
        assert_eq!(result[1].branch, Some("feat/auth".to_string()));
    }

    #[test]
    fn test_parse_worktree_list_detached() {
        let output = "\
worktree /home/user/repo
HEAD abc1234567890
branch refs/heads/main

worktree /home/user/repo-detached
HEAD def5678901234
detached

";
        let result = parse_worktree_list(output);
        assert_eq!(result.len(), 2);
        assert_eq!(result[1].branch, None);
    }

    #[test]
    fn test_parse_worktree_list_locked() {
        let output = "\
worktree /home/user/repo
HEAD abc1234567890
branch refs/heads/main

worktree /home/user/repo-locked
HEAD def5678901234
branch refs/heads/feat/locked
locked AI agent working

";
        let result = parse_worktree_list(output);
        assert_eq!(result.len(), 2);
        assert!(result[1].is_locked);
        assert_eq!(
            result[1].lock_reason,
            Some("AI agent working".to_string())
        );
    }

    #[test]
    fn test_parse_worktree_list_empty() {
        let result = parse_worktree_list("");
        assert!(result.is_empty());
    }
}

File: src/lib/stores/worktree.ts
import { writable, derived, get } from 'svelte/store';
import { repoPath } from '$lib/stores/repo';
import { addToast, updateToast } from '$lib/stores/toast';
import { addOutput } from '$lib/stores/output';
import * as api from '$lib/api/worktree';
import type { WorktreeInfo } from '$lib/api/types';

// State
export const worktrees = writable<WorktreeInfo[]>([]);
export const worktreeLoading = writable(false);
export const worktreeError = writable<string | null>(null);
export const showAddWorktreeDialog = writable(false);

// Derived
export const linkedWorktrees = derived(worktrees, ($wt) =>
  $wt.filter((w) => !w.is_main)
);
export const worktreeCount = derived(worktrees, ($wt) => $wt.length);

// Actions

export async function refreshWorktrees(): Promise<void> {
  const path = get(repoPath);
  if (!path) return;

  worktreeLoading.set(true);
  worktreeError.set(null);

  try {
    const list = await api.listWorktrees(path);
    worktrees.set(list);
  } catch (e) {
    worktreeError.set(e instanceof Error ? e.message : String(e));
  } finally {
    worktreeLoading.set(false);
  }
}

export async function createWorktree(
  worktreePath: string,
  branch: string | null,
  newBranch: boolean,
): Promise<boolean> {
  const path = get(repoPath);
  if (!path) return false;

  try {
    const result = await api.addWorktree(path, worktreePath, branch, newBranch);
    addToast(`Created worktree "${result.worktree.name}" (${result.worktree.branch ?? 'detached'})`, 'success');
    await refreshWorktrees();
    return true;
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    worktreeError.set(msg);
    addToast(`Failed to create worktree: ${msg}`, 'error');
    return false;
  }
}

export async function deleteWorktree(
  worktreePath: string,
  force: boolean,
): Promise<boolean> {
  const path = get(repoPath);
  if (!path) return false;

  try {
    const result = await api.removeWorktree(path, worktreePath, force);
    const name = worktreePath.split('/').pop() ?? worktreePath;
    if (result.success) {
      addToast(`Removed worktree "${name}"`, 'success');
    } else {
      const detail = result.output.stderr || 'Unknown error';
      worktreeError.set(detail);
      addToast(`Failed to remove worktree "${name}": ${detail}`, 'error');
    }
    await refreshWorktrees();
    return result.success;
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    worktreeError.set(msg);
    addToast(`Failed to remove worktree: ${msg}`, 'error');
    return false;
  }
}

export async function toggleWorktreeLock(
  worktreePath: string,
  currentlyLocked: boolean,
  reason?: string,
): Promise<void> {
  const path = get(repoPath);
  if (!path) return;

  try {
    let list: WorktreeInfo[];
    if (currentlyLocked) {
      list = await api.unlockWorktree(path, worktreePath);
    } else {
      list = await api.lockWorktree(path, worktreePath, reason);
    }
    worktrees.set(list);
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    worktreeError.set(msg);
    const op = currentlyLocked ? 'unlock' : 'lock';
    addToast(`Failed to ${op} worktree: ${msg}`, 'error');
  }
}

export async function deleteAllWorktrees(force: boolean): Promise<boolean> {
  const path = get(repoPath);
  if (!path) return false;

  const linked = get(linkedWorktrees);
  if (linked.length === 0) return true;

  const total = linked.length;
  const toastId = addToast(`Removing worktrees (0/${total})...`, 'info', 0);

  let removed = 0;
  let failed = 0;
  const errors: string[] = [];

  for (const wt of linked) {
    const name = wt.path.split('/').pop() ?? wt.path;
    updateToast(toastId, `Removing "${name}" (${removed + failed + 1}/${total})...`);

    try {
      const result = await api.removeWorktree(path, wt.path, force);
      if (result.success) {
        removed++;
      } else {
        failed++;
        errors.push(`${name}: ${result.output.stderr || 'Unknown error'}`);
      }
    } catch (e) {
      failed++;
      errors.push(`${name}: ${e instanceof Error ? e.message : String(e)}`);
    }
    await refreshWorktrees();
  }

  if (failed === 0) {
    updateToast(toastId, `Removed all ${removed} worktree(s)`, 'success', 4000);
  } else {
    updateToast(toastId, `Removed ${removed}/${total} worktree(s), ${failed} failed`, 'error', 6000);
    addOutput('Remove all worktrees', `Removed ${removed}/${total}`, errors.join('\n'), false);
  }

  return failed === 0;
}

export async function pruneStaleWorktrees(): Promise<void> {
  const path = get(repoPath);
  if (!path) return;

  try {
    const result = await api.pruneWorktrees(path, false);
    if (result.pruned.length > 0) {
      addToast(`Pruned ${result.pruned.length} stale worktree(s)`, 'success');
    } else {
      addToast('No stale worktrees to prune', 'info');
    }
    await refreshWorktrees();
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    worktreeError.set(msg);
    addToast(`Failed to prune worktrees: ${msg}`, 'error');
  }
}
